File: api/tts.py
from http.server import BaseHTTPRequestHandler
import json
import subprocess
import base64
import tempfile
import os
import sys

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
    
    def do_POST(self):
        try:
            content_length = int(self.headers.get('Content-Length', 0))
            post_data = self.rfile.read(content_length)
            data = json.loads(post_data.decode('utf-8'))
            
            text = data.get('text', '')[:5000]
            voice = data.get('voice', 'ru-RU-SvetlanaNeural')
            rate = data.get('rate', '1')
            pitch = data.get('pitch', '1')
            
            # Преобразуем параметры
            rate_f = float(rate)
            rate_str = f"+{int((rate_f - 1) * 100)}%" if rate_f >= 1 else f"{int((rate_f - 1) * 100)}%"
            
            pitch_f = float(pitch)
            pitch_str = f"+{int((pitch_f - 1) * 10)}Hz" if pitch_f >= 1 else f"{int((pitch_f - 1) * 10)}Hz"
            
            # Временный файл
            with tempfile.NamedTemporaryFile(delete=False, suffix='.mp3') as tmp:
                tmp_path = tmp.name
            
            try:
                # ВАЖНО: Используем python -m edge_tts
                result = subprocess.run([
                    sys.executable, '-m', 'edge_tts',
                    '--text', text,
                    '--voice', voice,
                    '--rate', rate_str,
                    '--pitch', pitch_str,
                    '--write-media', tmp_path
                ], capture_output=True, text=True, timeout=30)
                
                if result.returncode != 0:
                    raise Exception(f"edge-tts failed: {result.stderr}")
                
                # Читаем аудио
                with open(tmp_path, 'rb') as f:
                    audio_data = f.read()
                
                if len(audio_data) == 0:
                    raise Exception("Generated audio file is empty")
                
                base64_audio = base64.b64encode(audio_data).decode('utf-8')
                
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                
                self.wfile.write(json.dumps({'audio': base64_audio}).encode())
                
            finally:
                if os.path.exists(tmp_path):
                    os.remove(tmp_path)
                    
        except Exception as e:
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({'error': str(e)}).encode())

File: api/tts.js
import { exec } from 'child_process';
import { promisify } from 'util';
import { readFile, unlink, writeFile } from 'fs/promises';
import { tmpdir } from 'os';
import { join } from 'path';

const execAsync = promisify(exec);

export default async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  let tmpFile = null;

  try {
    const { text, voice = 'ru-RU-SvetlanaNeural', rate = '1', pitch = '1' } = req.body;

    if (!text) {
      return res.status(400).json({ error: 'Text is required' });
    }

    const textToSpeak = text.substring(0, 5000);

    // Преобразуем параметры
    const rateF = parseFloat(rate);
    const rateStr = rateF >= 1 
      ? `+${Math.round((rateF - 1) * 100)}%` 
      : `${Math.round((rateF - 1) * 100)}%`;

    const pitchF = parseFloat(pitch);
    const pitchStr = pitchF >= 1 
      ? `+${Math.round((pitchF - 1) * 10)}Hz` 
      : `${Math.round((pitchF - 1) * 10)}Hz`;

    // Временный файл
    tmpFile = join(tmpdir(), `tts_${Date.now()}.mp3`);

    // Сохраняем текст во временный файл (безопаснее чем передавать в команду)
    const textFile = join(tmpdir(), `text_${Date.now()}.txt`);
    await writeFile(textFile, textToSpeak, 'utf-8');

    try {
      // Проверяем наличие edge-tts
      let command;
      try {
        await execAsync('which edge-tts');
        // edge-tts установлен как команда
        command = `edge-tts --file "${textFile}" --voice "${voice}" --rate "${rateStr}" --pitch "${pitchStr}" --write-media "${tmpFile}"`;
      } catch {
        // Используем python -m edge_tts
        command = `python3 -m edge_tts --file "${textFile}" --voice "${voice}" --rate "${rateStr}" --pitch "${pitchStr}" --write-media "${tmpFile}"`;
      }

      const { stdout, stderr } = await execAsync(command, { 
        timeout: 30000,
        maxBuffer: 10 * 1024 * 1024
      });

      // Удаляем текстовый файл
      await unlink(textFile);

      // Читаем аудио
      const audioBuffer = await readFile(tmpFile);
      
      if (audioBuffer.length === 0) {
        throw new Error('Generated audio file is empty');
      }

      const base64Audio = audioBuffer.toString('base64');

      // Удаляем временный файл
      await unlink(tmpFile);

      return res.status(200).json({ audio: base64Audio });

    } catch (cmdError) {
      // Удаляем временные файлы при ошибке
      try { await unlink(textFile); } catch {}
      try { await unlink(tmpFile); } catch {}
      
      throw new Error(`edge-tts execution failed: ${cmdError.message}\nStderr: ${cmdError.stderr}\nStdout: ${cmdError.stdout}`);
    }

  } catch (error) {
    console.error('TTS Error:', error);
    
    // Очищаем временные файлы
    if (tmpFile) {
      try { await unlink(tmpFile); } catch {}
    }

    return res.status(500).json({ 
      error: error.message,
      stack: process.env.NODE_ENV === 'development' ? error.stack : undefined
    });
  }
}